File: DuarteQuestions/DuarteQuestions/Service/QuestionService.cs
using DuarteQuestions.CQRS.Questions.Command.CreateQuestion;
using DuarteQuestions.CQRS.Questions.Command.DeleteQuestion;
using DuarteQuestions.CQRS.Questions.Command.UpdateQuestion;
using DuarteQuestions.CQRS.Questions.Query.GetQuestionById;
using DuarteQuestions.CQRS.Questions.Query.GetQuestionList;
using DuarteQuestions.CQRS.Questions.ViewModel;
using DuarteQuestions.Service.Interface;
using MediatR;

namespace DuarteQuestions.Service
{
    public class QuestionService : IQuestionService
    {
        private readonly IMediator _mediator;

        public QuestionService(IMediator mediator)
        {
            _mediator = mediator;
        }

        public async Task<bool> CreateQuestion(CreateQuestionCommand command)
        {
            try
            {
                return await _mediator.Send(command);
            }
            catch (Exception)
            {
                throw;
            }
        }

        public async Task<bool> UpdateQuestion(UpdateQuestionCommand command)
        {
            try
            {
                return await _mediator.Send(command);
            }
            catch (Exception)
            {
                throw;
            }
        }

        public async Task<bool> DeleteQuestion(int id)
        {
            try
            {
                return await _mediator.Send(new DeleteQuestionCommand()
                {
                    Id = id
                });
            }
            catch (Exception)
            {
                throw;
            }
        }

        public async Task<IEnumerable<QuestionViewModel>> GetQuestionList(GetQuestionListQuery query)
        {
            try
            {
                return await _mediator.Send(query);
            }
            catch (Exception)
            {
                throw;
            }
        }

        public async Task<QuestionViewModel> GetQuestionById(int id)
        {
            try
            {
                return await _mediator.Send(new GetQuestionByIdQuery()
                {
                    Id = id
                });
            }
            catch (Exception)
            {
                throw;
            }
        }
    }
}

File: DuarteQuestionsWebPanel/src/app/services/question.service.ts
import { CreateQuestionCommand, GetQuestionListQuery, QuestionViewModel, UpdateQuestionCommand } from '../DTOs/models/models';
import { EndPoints } from '../endpoints/endpoints/endpoints';
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable, share } from 'rxjs';
import { environment } from '../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class QuestionService {
  private baseApiUrl: string = environment.baseApiUrl;
  private http: HttpClient = inject(HttpClient);

  constructor() {}
  
  public createQuestion(command: CreateQuestionCommand): Observable<boolean> {
    return this.http.post<boolean>(this.baseApiUrl + "/" + EndPoints.API_QUESTION_CREATE, command).pipe(share());
  }

  public updateQuestion(command: UpdateQuestionCommand): Observable<boolean> {
    return this.http.put<boolean>(this.baseApiUrl + "/" + EndPoints.API_QUESTION_UPDATE, command).pipe(share());
  }

  public deleteQuestion(id: number): Observable<boolean> {
    return this.http.delete<boolean>(this.baseApiUrl + "/" + EndPoints.API_QUESTION_DELETE + "/" + id).pipe(share());
  }

  public getQuestionList(query: GetQuestionListQuery): Observable<QuestionViewModel[]> {
    return this.http.get<QuestionViewModel[]>(this.baseApiUrl + "/" + EndPoints.API_QUESTION_GET_LIST + "?Keyword=" + query.keyword + "&GetAll=" + query.getAll).pipe(share());
  }

  public getQuestionById(id: number): Observable<QuestionViewModel> {
    return this.http.get<QuestionViewModel>(this.baseApiUrl + "/" + EndPoints.API_QUESTION_GET + "/" + id).pipe(share());
  }

  public restoreAll(): Observable<boolean> {
    return this.http.get<boolean>(this.baseApiUrl + "/" + EndPoints.API_QUESTION_RESTORE_ALL).pipe(share());
  }
}
